<a href="https://colab.research.google.com/github/NathanGeng/DS-3001/blob/main/assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

In [37]:
! git clone https://www.github.com/NathanGeng/assignment2



Cloning into 'assignment2'...
remote: Enumerating objects: 36, done.
remote: Counting objects: 100% (12/12), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 36 (delta 8), reused 5 (delta 5), pack-reused 24
Receiving objects: 100% (36/36), 5.47 MiB | 28.28 MiB/s, done.
Resolving deltas: 100% (9/9), done.


**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  


> This paper describes what tidying data is and looks to fill gaps in research with regards to how cleaning data is done. Tidying data would mean establishing a standard for how data is framed/presented, with standardized rows, columns, and tables. All of this would allow for more efficient and effective data analysis.



  2. Read the introduction. What is the "tidy data standard" intended to accomplish?



> The standardization of tidying data seeks to make data more organized. At the time when the paper was written, datasets had a lot of variance in how variables and observations were presented, which made it so that researchers had to "brute force" the data and translating its contents so that it could be analyzed by certain tools. A standard model would make this process more streamlined and make researchers do less menial tasks and focus more on the interesting parts of data.



  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."



> The first sentence means that there can be many ways in which a dataset becomes messy - in order to become a tidy dataset and fall into the standard, there must be many aspects in which it is cleaned. The second sentence means that, due to the confusing nature of data, variables and observations and their relationship can change due to the context they are presented in.



  4. Read Section 2.2. How does Wickham define values, variables, and observations?


> Values: What a dataset is comprised of; they can usually numbers (quantitative) or strings (qualitative), and have a corresponding variable and observation

> Variables: all values associated with a particular attribute; spans across all observations (ex: time, height)

> Observations: all values associated for a single unit; contains all variables (ex: patient, day)



  5. How is "Tidy Data" defined in section 2.3?


> A tidy dataset has variables, observertations, and types that match up well in columns, rows, and tables. Each variable should have its own column, each observation should have its own row, and they should combine to form a table.



  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?


> Column headers are their own value, not titles for the variable; there are multilple variables in a single column; there are variables in both rows and columns; different types of an observation are stored within the same table; a single observation is stored across multiple tables. The data in table 4 is messy because the rows in the table are variables, and the headers of the variables are values themselves. Melting a dataset means turning columns into rows.



  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?


> Table 11 is messy because there are variables in both columns and rows, Table 12 fixes this by creating individual observations by date and having each variable correspond to a column.



  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?


  > The problem is that tidy data can only be as efficient as the tools are that analyze them, and can therefore have a ceiling to its potential benefit. The author hopes that future work in data wrangling can develop even more effective practices, and tools that improve alongside those practices.

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)

> The one way I could think to clean the Price variable would to get rid of the commas in numbers greater than 999. This is because, when performing operations of numeric variables, commas can be disruptive and are better to be removed in most cases. Also, when getting rid of the commas, it turned the Price variable into an object, so I had to convert it back to a numeric. There end up being 181 missing values, checked using the isnull() function.

2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.

> To clean the Type variable, I first wanted to see what exactly the variable consisted of, what types of shark attacks were listed. I decided that boat, boating, the boating typo, and watercraft could all be designated as watercraft incidents. Also, for the less common types such as "Unconfirmed", "Questionable", etc., I grouped them all under the "Invalid" typing. Finally, I got rid of the null values within the column (but 5 remain for some reason).

3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.

> I replaced the 9 values with null and, since 0 means no and 1 means yes, I decided to change the names of those values as well to make the data more simple to understand.

4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

> When examining the 'Imposed Sentence' variable, it can be seen that there is no imposed sentence associated with the sentence type '4'. It can be inferred that there is no imposed sentencing time for type 4, so a value of 0 is put for sentence type 4.

In [38]:
import pandas as pd
import numpy as np
import os

os.chdir('/content/assignment2/data/')

In [39]:
df_airbnb = pd.read_csv('airbnb_hw.csv')
price = 'Price'

# getting rid of commas in price values
df_airbnb[price].str.replace(',','')

# converting back to numeric
df_airbnb[price] = pd.to_numeric(df_airbnb[price], errors='coerce') # here I'm coercing the Price values to numeric

# number of missing values
sum(df_airbnb[price].isnull())

181

In [40]:
df_sharks = pd.read_csv('sharks.csv')
var_type = 'Type'

# checking unique labels under the Type variable
print(df_sharks[var_type].value_counts())

# grouping types together
df_sharks[var_type] = df_sharks[var_type].replace(['Boat','Boating','Boatomg'],'Watercraft')
df_sharks[var_type] = df_sharks[var_type].replace(['Questionable','Unconfirmed','Unverified', 'Under investigation'],'Invalid')

# getting rid of nulls
df_sharks[var_type] = df_sharks[var_type].dropna(axis = 0)

# checking number of nulls
sum(df_sharks[var_type].isnull())

Unprovoked             4716
Provoked                593
Invalid                 552
Sea Disaster            239
Watercraft              142
Boat                    109
Boating                  92
Questionable             10
Unconfirmed               1
Unverified                1
Under investigation       1
Boatomg                   1
Name: Type, dtype: int64


<ipython-input-40-a4583a341272>:1: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  df_sharks = pd.read_csv('sharks.csv')


5

In [41]:
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
df_pretrial = pd.read_csv(url)

released = 'WhetherDefendantWasReleasedPretrial'

#replacing 9 values with null values
df_pretrial[released] = df_pretrial[released].replace(9,np.nan)

# changing 0 and 1 to no and yes
df_pretrial[released].replace(0, 'no', inplace=True)
df_pretrial[released].replace(1, 'yes', inplace=True)

print(df_pretrial[released].value_counts())

<ipython-input-41-77fdff248f00>:2: DtypeWarning: Columns (1,4,5,7,79,80,81,82,83,84,108,163,164,165,166,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,270,271,272,273,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,301,302,303,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,

yes    19154
no      3801
Name: WhetherDefendantWasReleasedPretrial, dtype: int64


In [43]:
imposed_sentence = 'ImposedSentenceAllChargeInContactEvent'
sentence_type = 'SentenceTypeAllChargesAtConvictionInContactEvent'

print(df_pretrial[sentence_type].value_counts())

df_pretrial.loc[df_pretrial[sentence_type] == 4, imposed_sentence] = 0

df_pretrial.loc[1:20,('ImposedSentenceAllChargeInContactEvent','SentenceTypeAllChargesAtConvictionInContactEvent')]



4    8779
0    8720
1    4299
2     914
9     274
Name: SentenceTypeAllChargesAtConvictionInContactEvent, dtype: int64


ImposedSentenceAllChargeInContactEvent  \
1                                      60   
2                                      12   
3                        .985626283367556   
4                                       0   
5                                      12   
6                                      36   
7                                       6   
8                                      24   
9                                      12   
10                       5.91375770020534   
11                                      0   
12                                      0   
13                                    120   
14                                     24   
15                                      0   
16                                      0   
17                                     12   
18                                      0   
19                                     12   
20                                     12   

    SentenceTypeAllChargesAtConvictionInContactEvent  
1                                                  0  
2                                                  1  
3                                                  1  
4                                                  4  
5                                                  0  
6                                                  1  
7                                                  0  
8                                                  1  
9                                                  1  
10                                                 0  
11                                                 4  
12                                                 4  
13                                                 1  
14                                                 1  
15                                                 4  
16                                                 4  
17                                                 0  
18                                                 4  
19                                                 0  
20                                                 0

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

Information taken from the United States Census Bureau (https://www.census.gov/en.html)

1. How did the most recent US Census gather data on race?

> It was based on self-identification, so all participants in the census would report all race groups they considered themselves a part of.

2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?

> The government collects census data because it is indicative of the direction which the nation is going in terms of population, how it is distributed across states, and potential trends. In terms of politics and society, census data is important because growth and decline in state populations influence the distribution of representatives in Congress for states. Additionally, it is used to provide the government an idea of where funding is best utilized/needed. Low quality data would lead to an ineffective use of government funds, and an inaccurate representation of the US people in federal government.

3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?

> A concern about how the census was conducted is the historical overrepresentation of certain groups and underrepresentation of others. An overrepresented population, for example, is white adults, who are more likely to be aware of the census and geographically mobile - all indicators of being more likely to participate in filling out a census form. An underrepresented group would be Hispanics, who are less likely to participate due to finances, or a distrust/stigma towards government. Something that could be interpreted as a strength would be their imputation methods to fill in gaps within the dataset using known data values. In the future, a way to fully capture the diverse population could be to train up individuals can bridge the gap between government and underrepresented communities. For example, training more Latin-American workers to go door-to-door during census collection to make individuals feel more at ease.

4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.

> The census only asks about sex; two simple boxes labeled male and female. This is accurate, as sex relates to biological attributes of males and females (based on hormones, chromosome composition, and bodily anatomy). However, the census does not ask about gender, a more socially constructed concept. While there is no obligation or requirement to ask about gender, questions pertaining to it would definitely provide a robust dataset that is more representative of the population as a whole. However, due to the expansive nature of gender, I find it hard to imagine how a question on the census would manage to encapsulate the many different forms of gender one can identify with, forcing people who identify as a gender different from the listed options to check a box labeled "Other" at the bottom. I believe that there should be question pertaining to gender, but how effective it would be at representing all backgrounds remains the question.

5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?

> One major concern would be the privacy of the indivdual when it comes to their willingness to share those characteristics which the government. Even though discrimination based on someone's race, sexual orientation, religion, etc. is wrong, there are still ways it can still occur subliminally. For example, if there are missing values in census data regarding these characteristics, they might be imputed as the "standard" choice, leading to a unrepresentative dataset. Or, arguably worse, if the data is correctly filled out, the government can discrimintate against areas with populations that hold certain characteristics (ex: black communities and redlining). To prevent this, a good practice to adopt would be to prevent biases against these groups from occuring in the first place. Also, it is important for the process of data cleaning to be transparent and forthcoming, so any possible source of bias can be detected easily when examining how the data was manipulated.

6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

> Accuracy of the algorithm is important; if not, there is the risk of misrepresenting one's background and influencing the validity of the data. Assuming that the algorithm is 100% accurate, there still remains the issue of privacy. I am assuming that the purpose of the algorithm is to impute values for these characteristics because participants chose to not disclose them; therefore, the ability for the algorithm to determine characteristics of a participant regardless of their consent or not would be a violation of the privacy of the individual. In a sense, even though the algorithm was intended for good (aiming to fill gaps in data to increase its validity), it could still contribute to potential discriminatory practices like those listed in the previous question.

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.